### PR TITLE
feat: using SSA for finalizer and primary patch

### PR DESCRIPTION
**Pull request by csviri**

### CODE REVIEW

**Comment by csviri** [comment]
@metacosm this is the implementation so far, will add integration test and docs that explains stuff but might worth to take a look on it now, if everything makes sense

**Comment by csviri** [comment]
@metacosm this PR is ready for review, the docs updates are missing only, will add it soon

**Review by metacosm — CHANGES_REQUESTED** [review]

File: operator-framework-core/src/main/java/io/javaoperatorsdk/operator/processing/event/ReconciliationDispatcher.java [hunk 1]
```diff
@@ -316,27 +327,53 @@ private PostExecutionControl<P> handleCleanup(P resource,
     return postExecutionControl;
   }
 
+  @SuppressWarnings("unchecked")
+  private P addFinalizerWithSSA(P originalResource) {
+    log.debug(
+        "Adding finalizer (using SSA) for resource: {} version: {}",
+        getUID(originalResource), getVersion(originalResource));
+    try {
+      P resource = (P) originalResource.getClass().getConstructor().newInstance();
```
[the remark this answers is not in the document]
Comment (csviri) [reply operator-framework-core/src/main/java/io/javaoperatorsdk/operator/processing/event/ReconciliationDispatcher.java:334]: Hmm, good point, will try it with `GenericKubernetesResource`

File: operator-framework-core/src/main/java/io/javaoperatorsdk/operator/processing/event/EventProcessor.java [hunk 2]
```diff
@@ -245,17 +245,6 @@ synchronized void eventProcessingFinished(
       state.markProcessedMarkForDeletion();
       metrics.cleanupDoneFor(resourceID, metricsMetadata);
     } else {
-      postExecutionControl
```
[the remark this answers is not in the document]
Comment (csviri) [reply operator-framework-core/src/main/java/io/javaoperatorsdk/operator/processing/event/EventProcessor.java:248]: This works only if the patch is with optimistic locking, but now neither of the patches do. Mainly for default SSA it definitely should not be, but we should not press it probably for non-SSA either.

File: operator-framework-core/src/main/java/io/javaoperatorsdk/operator/processing/event/ReconciliationDispatcher.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (csviri) [reply operator-framework-core/src/main/java/io/javaoperatorsdk/operator/processing/event/ReconciliationDispatcher.java:334]: Took a look but all our API-s work with typed API. Added a better message. Maybe we could try to optimize in a separate PR?
Also would not worry to much about performance, this only happens once, when the resource is created, on on every reconciliation.

**Pull request merged**